controller/volume_controller.go

```diff
@@ -1141,10 +1142,120 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 		if oldState != v.Status.State {
 			vc.eventRecorder.Eventf(v, v1.EventTypeNormal, EventReasonAttached, "volume %v has been attached to %v", v.Name, v.Status.CurrentNodeID)
 		}
+
+		keepLocalReplicaPolicy, err := vc.ds.GetSettingAsBool(types.SettingNameKeepLocalReplicaOnSameNodeAsEngine)
```

Comment:
You can put this setting check into the following function, then add extra checks and move function `handleKeepLocalReplicaOnSameNodeAsEngine` into the function `syncVolume` rather than `reconcileVolumeState`.

Reply:
thanks. Fixed